src/Sulu/Bundle/AdminBundle/Admin/Routing/PreviewFormRouteBuilder.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\AdminBundle\Admin\Routing;
+
+class PreviewFormRouteBuilder implements PreviewFormRouteBuilderInterface
```

Comment:
I think the incomplete type system of PHP prevents the usage of inheritance here.

for example, the `FormRouteBuilderInterface` includes the signature `setTabTitle(string $tabTitle): self`. Because of the missing covariance in the type system we need to explicitly set the return type of the `setTabTitle` to `FormRouteBuilderInterface` in the `FormRouteBuilder` class. 

now, if the `PreviewFormRouteBuilder` class extends the `FormRouteBuilder` class, the return type  of the `setTabTitle` method is fixed to `FormRouteBuilderInterface` instead of `PreviewFormRouteBuilderInterface`.

Reply:
or do i miss something obvious here?

Reply:
Hm, sounds correct... That is really a pity :confused:

Reply:
Random thought: Would it maybe work if we omit the type hints and replace them with phpdocs?

Reply:
i guess that could work because the phpdoc is ignored during runtime and we are on a low phpstan level. but i am not sure if this is worth it to break the consistency with other interface.

Reply:
I've got another idea, that might work... What if we introduce an abstract class used by the `FormRouteBuilder` and `PreviewFormRouteBuilder` doing the route options logic, and then both concrete `RouteBuilder` only do something like this:

```php
public function setFormKey(string $formKey): PreviewFormRouteBuilderInterface
{
    parent::setFormKey($formKey);

    return $this;
}
```

I have already [tested it](https://3v4l.org/Z3A31) and it should work. Then we don't have to duplicate that logic. Sorry, but I am really worried about this, because when we change some options for the form route, then I am quite sure that the `PreviewFormRouteBuilder` will not be updated :slightly_frowning_face:

Reply:
Or maybe we could use a trait for that, instead of inheritance?

Reply:
i understand that concern. will implement an abstract class here and also in https://github.com/sulu/sulu/pull/4452 🙂 

traits will not work because the method in the trait needs to specify the correct return type in order to be accepted by the interface. Therefore we cannot use the same trait for both classes as their interfaces require different return types.

Reply:
As said, the trait would return nothing at all, only set the properties on the route. And `return $this` is done in every Route Builder, similar to the example I've tested above: https://3v4l.org/Z3A31(only that we would be using traits instead).

Traits would also have the advantage that you could reuse a `FormRouteBuilderTrait` and a `ListRouteBuilderTrait` in the #4452.

Reply:
@danrot implemented this using traits and added some comments regarding the implementation